Command palette command discovery (#4154)
Textualize/textual: docs/guide/command_palette.md

    This is a deliberate design decision taken to prevent a single broken `Provider` class from making the command palette unusable.
    Errors in command providers will be logged to the [console](./devtools.md).

### discover method

The [`discover`][textual.command.Provider.discover] method is responsible for providing results (or *discovery hits*) that should be shown to the user when the command palette input is empty;
this is to aid in command discoverability.

!!! note

    Because `discover` hits are shown the moment the command palette is opened, these should ideally be quick to generate;
    commands that might take time to generate are best left for `search` -- use `discover` to help the user easily find the most important commands.

> `discover` is similar to `search` but with the difference that no search value is passed to it, and it should *yield* instances of [`DiscoveryHit`][textual.command.DiscoveryHit].
> There is no matching and no match score is needed.
> The [`DiscoveryHit`][textual.command.DiscoveryHit] contains information about how the hit should be displayed, and an optional help string;
> discovery hits are sorted in ascending alphabetical order.
> It also contains a callback, which will be run if the user selects that command.

Reviewer: This paragraph doesn't flow very well but this is a subjective thing.

Your paragraph reads a bit like a list that's phrased weird, so maybe you could refactor into something similar to the suggestion below, or maybe just rewrite the prose to flow a little bit better.

```suggestion
`discover` is similar to `search` but with these differences:
 - `discover` accepts no parameters (instead of the search value);
 - `discover` yields instances of [`DiscoveryHit`][textual.command.DiscoveryHit]` (instead of instances of xxx);
 - discovery hits are sorted in ascending alphabetical order because there is no matching and no match score is generated.
Instances of [`DiscoveryHit`][textual.command.DiscoveryHit] contain information about how the hit should be displayed, an optional help string, and a callback which will be run if the user selects that command.
```

(The suggestion above is not to be taken at face value. It's just to illustrate roughly what I had in mind.)

Author: Yeah, I feel that's probably cleaner; I'll merge that and tidy it up.
